src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java

```diff
@@ -0,0 +1,148 @@
+     * 3. Prepare Content from {@link UploadGeoJSONRequestContent#getData()}
+     * 4. Upload content
+     * 5. Delete pipeline
+     * @param content {@link UploadGeoJSONRequestContent} derived from {@link UploadGeoJSONRequest}
+     * @param isIndexAlreadyExists confirms whether the uploader should create the new index or not
+     */
+    public void upload(
+        UploadGeoJSONRequestContent content,
+        boolean isIndexAlreadyExists,
+        ActionListener<AcknowledgedResponse> flowListener
+    ) {
+        // validate input
+        Objects.requireNonNull(flowListener, "listener cannot be null");
+        Objects.requireNonNull(content, "content cannot be null");
+
+        // initialize step listeners to chain steps
+        StepListener<Void> createIndexStep = new StepListener<>();
+        StepListener<String> createPipelineStep = new StepListener<>();
+        StepListener<Void> uploadFeaturesStep = new StepListener<>();
+        StepListener<Exception> deletePipelineStep = new StepListener<>();
+
+        if (isIndexAlreadyExists) {
+            createIndexStep.onResponse(null); // mark create index step as completed, to continue to next steps.
+        } else {
+            // create index
+            Map<String, String> fieldMap = new HashMap<>();
+            fieldMap.put(content.getFieldName(), content.getFieldType());
+            indexManager.create(content.getIndexName(), Collections.unmodifiableMap(fieldMap), createIndexStep);
+        }
+        // create a pipeline after creating index
+        createIndexStep.whenComplete(
+            notUsed -> pipelineManager.create(content.getFieldName(), createPipelineStep),
+            flowListener::onFailure
+        );
+
+        // index features as document after creating pipeline
+        createPipelineStep.whenComplete(
+            pipeline -> { upload(pipeline, content, uploadFeaturesStep); },
+            createPipelineFailedException -> { flowListener.onFailure(createPipelineFailedException); }
+        );
+
+        // delete pipeline
+        uploadFeaturesStep.whenComplete(notUsed -> {
+            String pipeline = createPipelineStep.result();
+            pipelineManager.delete(pipeline, deletePipelineStep, () -> null);
+        }, uploadFailed -> {
+            String pipeline = createPipelineStep.result();
+            pipelineManager.delete(pipeline, deletePipelineStep, () -> uploadFailed);
+        });
+
+        // set response or failure depending on previous steps status
+        deletePipelineStep.whenComplete(uploadFailedException -> {
+            if (uploadFailedException != null) {
+                throw uploadFailedException;
+            }
+            flowListener.onResponse(new AcknowledgedResponse(true));
+        }, flowListener::onFailure);
+
+    }
+
+    private void upload(String pipeline, UploadGeoJSONRequestContent content, StepListener<Void> uploadStepListener) {
```

Comment:
I would probably call this index. upload seems to refer to the 5 step process, while this appears to just be focused on indexing.

Reply:
Ack